make sure that only concat preallocates buffers

ritchie46 opened the pull request:
# Which issue does this PR close?

Partial fix for #347. In #348  @jorgecarleitao pointed out that the memory savings don't work for the `filter` and the `zip` kernel. 
This PR restores the implementation for those two kernels, and keeps the new `preallocation` for the `concat` kernel. 

This is achieved by create a builder pattern for `MutableArrayData`. This way we don't break the API, and we may choose to `preallocate` buffers.

Could you take a look at this @jorgecarleitao ?

# Are there any user-facing changes?

There is an builder pattern struct for `MutableArrayData`.

# If there are any breaking changes to public APIs, please add the `breaking change` label.

No

jorgecarleitao commented:
Thanks @ritchie46 . My small concerns with this PR:

1. We are introducing yet another builder API for something relatively easy to accomplish with an extra method
2. `preallocate_buffers` assumes that "preallocate" means allocate the maximum possible capacity, which IMO is not really pre-allocating, is just being very conservative and allocating the maximum.
3. The API is rather limited to the `concatenate` case.

I left some ideas as a comment on the issue

ritchie46 commented:
@jorgecarleitao I implemented your proposal from #347. As we now need to define an `enum` in the `with_capacities`  constructor, I want to be extra certain that I added all possible variants, as adding one later is backwards incompatable. Could you check if I missed something.

For now, only the `(large)-utf8` ones uses this, but if the design is ok, I can add more in later PRs.

jhorstmann commented on arrow/src/array/transform/mod.rs:
```diff
@@ -338,33 +338,61 @@ fn preallocate_str_buffer<Offset: StringOffsetSizeTrait>(
+    /// * the capacities of the fields
+    Struct(usize, Option<Vec<Capacities>>),
+    /// Dictionary type
+    /// * the capacity of the array
```
What is the difference here between the "capacity of the array" and "the capacity of the keys"?

I have a rather special usage of concat for dictionary arrays, where I can guarantee that the dictionary values of all arrays are the same, or at least that the dictionary of the last array includes all previous dictionaries in the same order. I wonder whether that could be supported here and also fix the current concatenation logic that could lead to duplicated dictionary values. The rough idea would be to have another field here with an
```
enum DictionaryMergeStrategy {
    Append(Option<usize>), // current logic, just append dictionary values with an optional capacity
    Merge(Option<usize>), // actually merge the dictionaries using a hashmap, to be imlemented in a followup
    UseLastValues, // use the values of the last dictionary array, might want to validate that this has a size at least as large as all other dictionaries so all keys are guaranteed to be in range
}
```

ritchie46 replied:
> What is the difference here between the "capacity of the array" and "the capacity of the keys"?

Yes, that can be the same.

I also understand the use of defining the merge strategy. Regarding  the API, should we make that available in a different method. Similar to `HashMap` having a method:

```rust
HashMap::with_capacity(...)
// and
HashMap::with_capacity_and_hasher(...)
```

We could do

```rust
MutableArrayData::with_capacies(...)
// and
MutableArrayData::with_capacities_and_merge_strategy(...)
```

jhorstmann replied:
I haven't thought this through fully yet. Maybe the caller should actually specify the merged dictionary values array instead of a strategy. And that, or specifying the correct strategy, requires a special concat kernel anyway, so it's basically the same situation as now. I think you do not need to consider this use case in this PR.

alamb commented:
@jorgecarleitao  what is the status of this one?

alamb commented:
Included in #411
